[Room][Compiler Processing] Originating element improvements

elihart opened the pull request:
## Proposed Changes
Right now the compiler processing library only exposes a way to set originating elements when generating a file with JavaPoet, and KotlinPoet should be supported for the case of creating kotlin files. Additionally when writing a file with KSP the incremental mode defaults to isolating with no way to specify that it should be aggregating.

This change:
  - Adds a new boolean parameter to XFiler `write` functions to optionally specify isolating vs aggregating behavior (defaults to isolating)
  - Adds a kotlinpoet extension function to add originating elements.
  - Exposes a public function to convert XElement to an originating element in case users need to access it directly

## Testing

Test: OriginatingElementsTest

Unit tests that check originating elements are created from XElement and added to kotlin and java poet type builders.

## Issues Fixed

Fixes: https://issuetracker.google.com/issues/184568922

elihart committed:
[XProcessor] Add aggregating and kotlinpoet extensions

elihart committed:
add tests

elihart commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/JavaPoetExt.kt:
```diff
@@ -51,12 +51,9 @@ internal fun TypeMirror.safeTypeName(): TypeName = if (kind == TypeKind.NONE) {
  * see [TypeSpec.Builder.addOriginatingElement].
  */
-fun TypeSpec.Builder.addOriginatingElement(element: XElement) {
-    if (element is JavacElement) {
```
Pulled this implementation out to share with the kotlin poet extension, and to make it public in case any one wants their own custom file writing.

I also changed this to return the builder, which is the pattern that javapoet uses and allows better chaining

elihart commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KSFileAsOriginatingElement.kt:
```diff
@@ -30,7 +30,7 @@ import javax.lang.model.type.TypeMirror
  * get the actual KSFile out of it.
  */
-internal class KSFileAsOriginatingElement(
+internal data class KSFileAsOriginatingElement(
```
Made this a data class to support equality for testing purposes

yigit replied:
sgtm

elihart committed:
fix lint issues

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XElement.kt:
```diff
@@ -75,3 +78,18 @@ fun XElement.isConstructor(): Boolean {
+ * Note, instead of using this directly you can instead use the [addOriginatingElement] extension
+ * functions for JavaPoet and KotlinPoet Type builders.
+ */
+fun XElement.originatingElement(): Element? {
```
So far, I don't think we ever exposed JavaAP types from the API.
Maybe, it is better to have a more explicit name not to get confused with XElement? Maybe something like:

`originatingJavaAPElement`
or
`originatingElementForJavaPoet`
or
`originatingElementForPoet` to also cover for KotlinPoet?


might also make sense to move this to a real method in XElement interface so that if we ever have another type of it, we won't forget to implement this. But I'm fine with this implementation as well, just an idea.

elihart replied:
sure, I can move `originatingElementForPoet` to the interface

elihart replied:
actually I realized that publicly exposing this isn't actually that helpful, since if someone did want to use it for some custom writing they wouldn't be able to do so in KSP since they can't access the `KSFileAsOriginatingElement`. We would need to provide a more thought out abstraction for that, which seems overkill for now since it is unlikely someone is using neither javapoet or kotlinpoet, so I'll just go back to having this be internal

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XFiler.kt:
```diff
@@ -23,11 +23,22 @@ import com.squareup.kotlinpoet.FileSpec
 interface XFiler {
-    fun write(javaFile: JavaFile)
+    /**
+     * @param aggregating Specifies whether this
```
lets make this an enum (can still have the default).
Boolean makes it harder to read if not called w/ named parameters.

yigit reviewed (commented):
overall lgtm, just minor nits.

elihart committed:
Remove public visibility of originating element

elihart committed:
Add enum based filing mode

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XElement.kt:
```diff
@@ -75,3 +79,19 @@ fun XElement.isConstructor(): Boolean {
+ * that allows us to pass originating elements to JavaPoet and KotlinPoet, and later extract
+ * the KSP file when writing with [XFiler].
+ */
+internal fun XElement.originatingElementForPoet(): Element? {
```
oh was this always internal? i thought it was going to be public. anyways, i prefer internal unless we have to make it public (mostly because KSFileAsOriginatingElement is not a true implementation)

elihart replied:
you're right, it was public but I changed it back to internal since I had the same thought https://github.com/androidx/androidx/pull/155#discussion_r608116580

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/XElement.kt:
```diff
@@ -75,3 +79,19 @@ fun XElement.isConstructor(): Boolean {
+    return when (this) {
+        is JavacElement -> element
+        is KspElement -> containingFileAsOriginatingElement()
+        else -> error("Originating element not implemented for ${this.javaClass}")
```
nit: is not implemented

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/javac/JavacFiler.kt:
```diff
@@ -22,11 +22,13 @@ import com.squareup.kotlinpoet.FileSpec
-    override fun write(javaFile: JavaFile) {
+
+    // "aggregating" is ignored in javac, and only applicable in KSP
+    override fun write(javaFile: JavaFile, aggregating: Boolean) {
```
forgot to push this change? i saw you made it an enum above but it is still a boolean here?

elihart replied:
Oh, I forgot to update the implementations

yigit commented on room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KspFiler.kt:
```diff
@@ -32,23 +32,24 @@ internal class KspFiler(
     private val messager: XMessager,
 ) : XFiler {
-    override fun write(javaFile: JavaFile) {
+    override fun write(javaFile: JavaFile, aggregating: Boolean) {
```
ditto

elihart committed:
update xfiler implementations with enum

yigit commented:
The build failed with: (in `AutoMigrationWriterTest`)

```
Caused by: java.lang.IllegalStateException: Originating element is not implemented for class androidx.room.compiler.processing.XElement$MockitoMock$1838500372
	at androidx.room.compiler.processing.XElementKt.originatingElementForPoet(XElement.kt:95)
	at androidx.room.compiler.processing.JavaPoetExtKt.addOriginatingElement(JavaPoetExt.kt:53)
	at androidx.room.writer.AutoMigrationWriter.createTypeSpecBuilder(AutoMigrationWriter.kt:50)
	at androidx.room.writer.ClassWriter.write(ClassWriter.kt:68)
	at androidx.room.writer.AutoMigrationWriterTest$validAutoMigrationWithDefaultValue$1.invoke(AutoMigrationWriterTest.kt:77)
	at androidx.room.writer.AutoMigrationWriterTest$validAutoMigrationWithDefaultValue$1.invoke(AutoMigrationWriterTest.kt:32)
```

you should be able to repro with a rebase. I thought your change wouldn't change the behavior so maybe i missed that.

yigit commented:
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:room/compiler/src/test/kotlin/androidx/room/writer/AutoMigrationWriterTest.kt;l=76?q=AutoMigrationWriterTest

elihart committed:
fix tests in AutoMigrationWriterTest

elihart commented:
Fixed. The tests started failing because I changed the originating element function throw an error if an unknown type is encountered, which failed for a mocked XElement - fixed by using a real implementation in the test

elihart commented:
Navigation lint seems to be failing for unrelated reasons, not sure I can do anything about it besides try rebasing later and hope its fixed upstream

```
/home/runner/work/androidx/androidx/navigation/navigation-dynamic-features-fragment/src/main/java/androidx/navigation/dynamicfeatures/fragment/ui/AbstractProgressFragment.kt:70: Error: Upgrade Fragment version to at least 1.3.0. [InvalidFragmentVersionForActivityResult]
    private val intentSenderLauncher = registerForActivityResult(
    ```
